### PR TITLE
etcdserver: internal request union

### PR DESCRIPTION
**Pull request by xiang90**
prepare for v3 api request

The wire types of Request and InternalRaftRequest are different. So we do not need the magic number.

/cc @yichengq @barakmich @philips

### CODE REVIEW

**Comment by philips** [comment]
I would rather see us move to a new API using an upgrade mechanism driven through consensus instead of hacking around this with guessing. Essentially I think that we need to make the Metadata proto look like 

```
message Metadata {
	required uint64 NodeID    = 1 [(gogoproto.nullable) = false];
	required uint64 ClusterID = 2 [(gogoproto.nullable) = false];
	optional uint64 CurrentVersion = 3 [(gogoproto.nullable) = false];
	optional uint64 MaxVersion = 4 [(gogoproto.nullable) = false];
}
```

Then at some point when all members have MaxVersion that is greater than CurrentVersion we move up to the largest agreed upon MaxVersion of the internal protocol. That way we can gate things like say the ACL stuff being available to the client API until all members support that protocol.

We talked about this before in Documentation/internal-protocol-versioning.md but that doesn't apply anymore and the file should probably be removed.

**Comment by xiang90** [comment]
@philips I think we are talking about different things here. 

The first thing here is to  make etcdserver be able to understand potentially unlimited proto message. 

The approach I took here is to use a union.  What to be included in the outer proto is another thing needed to be discussed.

**Comment by xiang90** [comment]
The version coordination is a separate thing. And the metadata proto is not used in the wire message at all. It is used by WAL to identify the local member/cluster right now.

**Comment by xiang90** [comment]
@philips @barakmich @yichengq Can you please take a look? We want to add v3 request proto to the union.

File: etcdserver/server.go [hunk 1]
```diff
@@ -745,9 +747,19 @@ func (s *EtcdServer) apply(es []raftpb.Entry, confState *raftpb.ConfState) (uint
 				}
 				break
 			}
-			var r pb.Request
-			pbutil.MustUnmarshal(&r, e.Data)
-			s.w.Trigger(r.ID, s.applyRequest(r))
+
+			var raftReq pb.InternalRaftRequest
+			if !pbutil.MaybeUnmarshal(&raftReq, e.Data) { // backward compatible
```
Comment (yichengq) [inline etcdserver/server.go]: Is it possible that one InternalRaftRequest is unmarshaled to Request?

Comment (xiang90) [inline etcdserver/server.go]: they have different wire type. so no.

Comment (xiang90) [inline etcdserver/server.go]: i can write test for it

Comment (yichengq) [inline etcdserver/server.go]: Ack. If we add requestv3 into InternalRaftRequest, we should be careful that its wire type is different from related tag in Request.

Maybe we could add a conflict wire type field into InternalRaftRequest. But it is totally ugly.

Comment (xiang90) [inline etcdserver/server.go]: i actually cannot image how will we manage to make the wire type the same... so i do not have the concern right now.

**Comment by xiang90** [comment]
@yichengq I am going to merge this if there is no objection.

**Comment by yichengq** [comment]
lgtm

**Pull request merged**